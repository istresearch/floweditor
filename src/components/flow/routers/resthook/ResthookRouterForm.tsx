import { react as bindCallbacks } from 'auto-bind';
import Dialog, { ButtonSet } from 'components/dialog/Dialog';
import { hasErrors, renderIssues } from 'components/flow/actions/helpers';
import { RouterFormProps } from 'components/flow/props';
import { createResultNameInput } from 'components/flow/routers/widgets';
import AssetSelector from 'components/form/assetselector/AssetSelector';
import TypeList from 'components/nodeeditor/TypeList';
import * as React from 'react';
import { FormEntry, FormState, mergeForm, StringEntry } from 'store/nodeEditor';
import {
  Alphanumeric,
  Required,
  shouldRequireIf,
  StartIsNonNumeric,
  validate
} from 'store/validators';

import { nodeToState, stateToNode } from './helpers';
import styles from './ResthookRouterForm.module.scss';
import i18n from 'config/i18n';
import { SelectOption } from 'components/form/select/SelectElement';

// TODO: Remove use of Function
export interface ResthookRouterFormState extends FormState {
  resthook: FormEntry;
  resultName: StringEntry;
}

export default class ResthookRouterForm extends React.PureComponent<
  RouterFormProps,
  ResthookRouterFormState
> {
  options: SelectOption[] = [];
  constructor(props: RouterFormProps) {
    super(props);

    this.state = nodeToState(props.nodeSettings);

    bindCallbacks(this, {
      include: [/^on/, /^handle/]
    });
  }

  public componentDidMount(): void {
    const items = this.props.assetStore.resthooks ? this.props.assetStore.resthooks.items : {};
    this.options = Object.keys(items).map((key: string) => {
      return { name: items[key].name, value: key };
    });
  }

  private handleUpdateResultName(result: string): void {
    const resultName = validate(i18n.t('forms.result_name', 'Result Name'), result, [
      Required,
      Alphanumeric,
      StartIsNonNumeric
    ]);
    this.setState({
      resultName,
      valid: this.state.valid && !hasErrors(resultName)
    });
  }

  public handleResthookChanged(selected: any[], submitting = false): boolean {
    const updates: Partial<ResthookRouterFormState> = {
      resthook: validate(i18n.t('forms.resthook', 'Resthook'), selected[0], [
        shouldRequireIf(submitting)
      ])
    };

    const updated = mergeForm(this.state, updates);
    this.setState(updated);
    return updated.valid;
  }

  private handleSave(): void {
    // validate our resthook in case they haven't interacted
    const valid = this.handleResthookChanged([this.state.resthook.value], true);

    if (valid) {
      this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
      this.props.onClose(false);
    }
  }

  public getButtons(): ButtonSet {
    return {
      primary: { name: 'Ok', onClick: this.handleSave },
      secondary: { name: 'Cancel', onClick: () => this.props.onClose(true) }
    };
  }

  public render(): JSX.Element {
    const typeConfig = this.props.typeConfig;
    return (
      <Dialog title={typeConfig.name} headerClass={typeConfig.type} buttons={this.getButtons()}>
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />
        <AssetSelector
          name={i18n.t('forms.resthook', 'Resthook')}
          placeholder={i18n.t('forms.resthook_to_call', 'Select the resthook to call')}
          assets={this.props.assetStore.resthooks}
          entry={this.state.resthook}
          searchable={true}
          onChange={this.handleResthookChanged}
          nameKey="resthook"
          valueKey="resthook"
        />
        <div className={styles.result_name}>
          {createResultNameInput(this.state.resultName, this.handleUpdateResultName)}
        </div>
        {renderIssues(this.props)}
      </Dialog>
    );
  }
}
