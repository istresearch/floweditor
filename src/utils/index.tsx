import {
  Action,
  Case,
  Category,
  ContactProperties,
  FlowPosition,
  LocalizationMap
} from 'flowTypes';
import { Query } from 'immutability-helper';
import * as React from 'react';
import Localization, { LocalizedObject } from 'services/Localization';
import { Asset, Assets } from 'store/flowContext';
import { FormEntry } from 'store/nodeEditor';
import { v4 as generateUUID } from 'uuid';
import variables from 'variables.module.scss';

export const V4_UUID = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;
const LABEL_CHARS = /^[a-zA-Z0-9-][a-zA-Z0-9- ]*$/;
const MAX_LABEL_LENGTH = 36;
const SNAKED_CHARS = /\s|-+(?=\S)/g;

export const MAX_REFLOW_ATTEMPTS = 100;
export const ACTIVITY_INTERVAL = 5000;
export const COLLISION_FUDGE = 5;
export const GRID_SIZE: number = parseInt(variables.grid_size, 10) || 20;
export const NODE_SPACING: number = parseInt(variables.node_spacing, 10) || 10;
export const NODE_PADDING: number = parseInt(variables.node_padding, 10) || 10;
export const QUIET_NOTE = 2000;
export const CONFIRMATION_TIME = 2000;

export interface BoolMap {
  [key: string]: boolean;
}

interface Bounds {
  left: number;
  top: number;
  right: number;
  bottom: number;
}

export type LabelIdCb = (label?: string, labelKey?: string, valueKey?: string) => string;

/**
 * Adjusts the left and top offsets to a grid
 * @param left horizontal offset
 * @param top vertical offset
 */
export const snapToGrid = (left: number, top: number): { left: number; top: number } => {
  let leftAdjust = left % GRID_SIZE;
  let topAdjust = top % GRID_SIZE;

  if (leftAdjust > GRID_SIZE / 3) {
    leftAdjust = GRID_SIZE - leftAdjust;
  } else {
    leftAdjust = leftAdjust * -1;
  }

  if (topAdjust > GRID_SIZE / 3) {
    topAdjust = GRID_SIZE - topAdjust;
  } else {
    topAdjust = topAdjust * -1;
  }

  return {
    left: Math.max(left + leftAdjust, 0),
    top: Math.max(top + topAdjust, 0)
  };
};

/**
 * Adjusts the position offsets to a grid
 */
export const snapPositionToGrid = (position: FlowPosition): FlowPosition => {
  let leftAdjust = position.left % GRID_SIZE;
  let topAdjust = position.top % GRID_SIZE;

  if (leftAdjust > GRID_SIZE / 3) {
    leftAdjust = GRID_SIZE - leftAdjust;
  } else {
    leftAdjust = leftAdjust * -1;
  }

  if (topAdjust > GRID_SIZE / 3) {
    topAdjust = GRID_SIZE - topAdjust;
  } else {
    topAdjust = topAdjust * -1;
  }

  const left = Math.max(position.left + leftAdjust, 0);
  const top = Math.max(position.top + topAdjust, 0);
  const right = left + position.right - position.left;
  const bottom = top + position.bottom - position.top;

  return {
    left,
    top,
    right,
    bottom
  };
};

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {object} A map of each string
 */
export const toBoolMap = (array: string[]): BoolMap =>
  array.reduce(
    (map: BoolMap, item: string) => ({
      ...map,
      [item]: true
    }),
    {}
  );

/**
 * Inserts commas into numbers where appropriate for better readability
 * @param {number} value - A number, e.g. 10000
 * @returns {string} A comma-separated string, e.g. 10,000
 */
export const addCommas = (value: number): string =>
  value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export const snakify = (value: string): string =>
  value
    .toLowerCase()
    .trim()
    .replace(SNAKED_CHARS, '_');
/**
 * Returns true if a given UUID matches v4 format
 * @param {string} uuid - A version 4 UUID (no braces, uppercase OK)
 * @returns {boolean}
 */
export const validUUID = (uuid: string): boolean => V4_UUID.test(uuid);

/**
 * Returns a given string in title case, e.g. 'full name' becomes 'Full Name'
 * @param {string} str - string to be title-cased
 * @returns {string} Title-cased string
 */
export const titleCase = (str: string): string =>
  str.replace(/\b\w+/g, s => s.charAt(0).toUpperCase() + s.substr(1).toLowerCase());

export const getSelectClassForEntry = (entry: FormEntry): string => {
  if (entry && entry.validationFailures && entry.validationFailures.length > 0) {
    return 'select-invalid';
  }
  return '';
};

export const getSelectClass = (errors: number): string => {
  if (errors === 0) {
    return 'react-select select-base';
  }
  // We use a global selector here for react-select
  return 'react-select select-base select-invalid';
};

export const reorderList = (list: any[], startIndex: number, endIndex: number): any[] => {
  const [removed] = list.splice(startIndex, 1);

  list.splice(endIndex, 0, removed);

  return list;
};

/**
 * Compares basic objects (no methods and DOM nodes; property order important)
 * @param {object} objA - basic object
 * @param {object} objB - basic object
 * @returns {boolean}
 */
export const jsonEqual = (objA: {}, objB: {}): boolean =>
  JSON.stringify(objA) === JSON.stringify(objB);

/**
 * Checks whether any of a list of error strings contain one or more queries.
 * Used to determine whether a particular error has been encountered.
 */
export const hasErrorType = (errors: string[], exps: RegExp[]): boolean => {
  if (!errors.length) {
    return false;
  }
  for (const error of errors) {
    for (const exp of exps) {
      if (error.match(exp)) {
        return true;
      }
    }
  }
  return false;
};

export const getLocalizedObject = (localizations: LocalizedObject[]) => {
  if (localizations && localizations.length) {
    return localizations[0];
  }
};

export interface ClickHandler {
  // onClick: (event: React.MouseEvent<HTMLDivElement>) => void;
  onMouseDown: (event: React.MouseEvent<HTMLDivElement>) => void;
  onMouseUp: (event: React.MouseEvent<HTMLDivElement>) => void;
}

/**
 * Creates a simple click handler via onMouseDown and onMouseUp.
 * This is a necessity in order to let jsPlumb manage our element dragging.
 * @param onClick
 */
export const createClickHandler = (
  onClick: (event: React.MouseEvent<HTMLElement>) => void,
  shouldCancelClick: () => boolean = null,
  onMouseDown?: (event: React.MouseEvent<HTMLElement>) => void
): any => {
  return {
    onMouseDown: (event: React.MouseEvent<HTMLElement>) => {
      if (onMouseDown) {
        onMouseDown(event);
      }
    },
    onMouseUp: (event: React.MouseEvent<HTMLElement>) => {
      if (!shouldCancelClick || !shouldCancelClick()) {
        onClick(event);
      }
    }
  };
};

export const getLocalization = (
  obj: Action | Category | Case,
  localization: LocalizationMap,
  language: Asset
) => Localization.translate(obj, language, localization[language.id]);

/** istanbul ignore next */
export const dump = (thing: any) => console.log(JSON.stringify(thing, null, 4));

/**
 * Apply emphasis style
 */
export const emphasize = (text: string) => <span className="emphasize">{text}</span>;
export const ellipsize = (text: string, max: number = 250) => {
  if (text.length > max) {
    return text.substr(0, max) + '...';
  }
  return text;
};

/**
 * Does property exist in our ContactProperties enum?
 * @param {string} propertyToCheck - 'name' key on react-select AutoCompleteOption
 */
export const propertyExists = (propertyToCheck: string) => {
  for (const property of Object.keys(ContactProperties)) {
    if (property.toLowerCase() === propertyToCheck.toLowerCase().trim()) {
      return true;
    }
  }
  return false;
};

/**
 * Should x element be rendered?
 */
export const renderIf = (predicate: boolean) => (then: JSX.Element, otherwise?: JSX.Element) =>
  predicate ? then : otherwise ? otherwise : null;

/**
 * Does the label meet our length requirements?
 * @param {string} label - label created by react-select
 */
export const properLabelLength = (label: string = '') =>
  label.length > 0 && label.length <= MAX_LABEL_LENGTH;

/**
 * Does the label meet our character requirements?
 * @param {string} label
 */
export const containsOnlyLabelChars = (label: string = '') => LABEL_CHARS.test(label);

/**
 * Does the label meet our length, character requirements?
 * @param {string} label - label created by react-select
 */
export const isValidLabel = (label: string) =>
  properLabelLength(label) && containsOnlyLabelChars(label);

export const isRealValue = (obj: any) => obj !== null && obj !== undefined;

/* istanbul ignore next */
export const timeStart = (name: string) =>
  process.env.NODE_ENV === 'development' && console.time(name);

/* istanbul ignore next */
export const timeEnd = (name: string) =>
  process.env.NODE_ENV === 'development' && console.timeEnd(name);

export const log = (...message: any[]) =>
  process.env.NODE_ENV === 'development' && console.log(...message);

export const capitalize = (str: string) =>
  str.replace(/(?:^|\s)\S/g, captured => captured.toUpperCase());

export const set = (val: any): Query<any> => ({ $set: val });

export const setTrue = (): Query<true> => set(true);

export const setFalse = (): Query<false> => set(false);

export const setNull = (): Query<null> => set(null);

export const setEmpty = (): Query<string> => set('');

export const merge = (val: any): Query<any> => ({ $merge: val });

export const unset = (val: any): Query<any> => ({ $unset: val });

export const push = (arr: any[]): Query<any[]> => ({ $push: arr });

// tslint:disable-next-line:array-type
export const splice = (arr: Array<Array<any>>): Query<Array<Array<any>>> => ({
  $splice: arr
});

export const optionExists = (newOptName: string, options: any[]) =>
  options.find(({ name }) => name.toLowerCase().trim() === newOptName.toLowerCase().trim())
    ? true
    : false;

/* export const isOptionUnique: IsOptionUniqueHandler = ({ option, options, labelKey, valueKey }) =>
    !propertyExists(option.name) && !optionExists(option.name, options);

export const isValidNewOption: IsValidNewOptionHandler = ({ label }) => isValidLabel(label);
*/
/* export const composeCreateNewOption = ({
    idCb,
    type
}: {
    idCb: LabelIdCb;
    type: AssetType;
}): NewOptionCreatorHandler => ({ label, labelKey, valueKey }) => ({
    id: idCb(label, labelKey, valueKey),
    name: label,
    type,
    isNew: true
});*/

/**
 * Deduplicate values in an array.
 * Pass it a key to deduplicate based on obj key.
 */
export const uniqueBy = (a: any[], key: string): any[] => {
  const seen: any = {};
  return a.filter((item: any) => {
    const k = item[key];
    return seen.hasOwnProperty(k) ? false : (seen[k] = true);
  });
};

export const downloadJSON = (obj: any, name: string): void => {
  const dataStr =
    'data:text/json;charset=utf-8,' + encodeURIComponent(JSON.stringify(obj, null, 2));
  const downloadAnchorNode = document.createElement('a');
  downloadAnchorNode.setAttribute('href', dataStr);
  downloadAnchorNode.setAttribute('download', name + '.json');
  document.body.appendChild(downloadAnchorNode); // required for firefox
  downloadAnchorNode.click();
  downloadAnchorNode.remove();
};

export const fetchAsset = (asset: Asset) => (assets: Assets, id: string): Promise<Asset> => {
  return new Promise<Asset>((resolve, reject) => {
    resolve(asset);
  });
};

export let createUUID = (): string => {
  return generateUUID();
};

export const seededUUIDs = (seed: number = 1): any => {
  let current = seed;
  const randomNext = () => {
    const x = Math.sin(current++) * 10000;
    return x - Math.floor(x);
  };

  const random = (min: number, max: number): number => {
    return Math.floor(randomNext() * (max - min) + min);
  };

  return () => {
    const randomSeed: number[] = [];
    for (let idx = 0; idx < 16; idx++) {
      randomSeed.push(random(0, 250));
    }
    return generateUUID({ random: randomSeed });
  };
};

export const range = (start: number, end: number) =>
  Array.from({ length: end - start }, (v: number, k: number) => k + start);

export const pluralize = (count: number, noun: string, suffix: string = 's'): string =>
  `${noun}${count !== 1 ? suffix : ''}`;

export const hasString = (names: string[], key: string): boolean =>
  !!names.find((item: string) => item === key);

export const scalarArrayEquals = (a: any[], b: any[]) =>
  a.every((value, index) => value === b[index]);

export const getURNPath = (urn: string) => {
  return urn.split(':')[1];
};

export const copyToClipboard = (text: string) => {
  navigator.clipboard.writeText(text);
};

export const throttle = (func: any, timeout: any) => {
  let ready: boolean = true;

  return (...args: any) => {
    if (!ready) {
      return;
    }

    ready = false;
    func(...args);
    setTimeout(() => {
      ready = true;
    }, timeout);
  };
};

export const traceUpdate = (component: any, prevProps: any, prevState?: any) => {
  const messages: string[] = [];
  Object.entries(component.props).forEach(
    ([key, val]) => prevProps[key] !== val && messages.push(`Prop: '${key}' changed`)
  );
  if (prevState && component.state) {
    Object.entries(component.state).forEach(
      ([key, val]) => prevState[key] !== val && messages.push(`State: '${key}' changed`)
    );
  }

  if (messages.length > 0) {
    log('****  ' + component._reactInternalFiber.type.name + ' changed');
    messages.forEach((message: string) => {
      log(message);
    });
  }
};

export const debounce = (fn: any, quiet: number, closure: any = null) => {
  if (fn.timeout) {
    window.clearTimeout(fn.timeout);
  }
  fn.timeout = window.setTimeout(closure || fn, quiet);
};

export const onNextRender = (fn: any) => {
  window.setTimeout(fn, 0);
};

export const desnake = (text: string): string => {
  if (!text) {
    return text;
  }
  return text
    .split('_')
    .join(' ')
    .toLowerCase();
};

export const bool = (prop: boolean) => {
  return prop ? 'true' : null;
};
