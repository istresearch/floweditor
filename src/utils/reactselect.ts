import { StylesConfig } from 'react-select';

export const large = {
  placeholder: (styles: StylesConfig, state: any) => {
    return { ...styles, color: '#ddd', marginLeft: '3px' };
  },
  singleValue: (styles: StylesConfig, state: any) => {
    return { ...styles, marginLeft: '3px' };
  },
  multiValue: (styles: StylesConfig, state: any) => {
    return { ...styles, marginLeft: '0px' };
  },
  input: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      marginLeft: '0px',
      caretColor: '#999',
      marginBottom: '0px',
      boxShadow: 'none',
      border: 'none',
      fontSize: '13px'
    };
  },
  control: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      borderColor: '#e6e6e6',
      boxShadow: 'none',
      fontSize: '13px',
      transition: 'all 0.3s ease-in-out',
      '&:hover': {
        borderColor: '#e6e6e6'
      },
      '&:focus-within': {
        boxShadow: 'var(--widget-box-shadow-focused) !important',
        border: '1px solid #a4cafe !important'
      }
    };
  }
};

export const messageStyle = {
  control: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      borderColor: '#e6e6e6',
      boxShadow: 'none',
      borderBottom: 'none',
      borderBottomRightRadius: '0px',
      borderBottomLeftRadius: '0px',
      '&:hover': {
        borderColor: '#e6e6e6'
      }
    };
  }
};

export const getErroredSelect = (baseControl: any): any => {
  return {
    control: (styles: StylesConfig) => ({
      ...styles,
      ...baseControl,
      borderColor: 'tomato',
      boxShadow: '0 0 0 3px rgba(255,196,186,0.5) !important;',
      '&:hover': {
        borderColor: 'tomato'
      }
    })
  };
};

export const small = {
  control: (styles: StylesConfig) => ({
    ...styles,
    borderColor: '#e6e6e6',
    boxShadow: 'none',
    '&:hover': {
      borderColor: '#e6e6e6'
    },
    maxHeight: '28px',
    minHeight: '28px',
    fontSize: '12px',
    lineHeight: '12px',
    marginBottom: '-10px',
    '&:focus-within': {
      boxShadow: 'var(--widget-box-shadow-focused) !important',
      border: '1px solid #a4cafe !important'
    }
  }),
  menuList: (styles: StylesConfig) => ({
    ...styles,
    maxHeight: 200
  }),
  option: (styles: StylesConfig) => ({
    ...styles,
    fontSize: 12
  }),
  singleValue: (styles: StylesConfig) => ({
    ...styles,
    maxHeight: 28,
    lineHeight: '13px'
  }),
  indicatorsContainer: (styles: StylesConfig) => ({
    ...styles,
    height: 28,
    maxHeight: 28
  }),
  valueContainer: (styles: StylesConfig) => ({
    ...styles,
    height: 28,
    maxHeight: 28
  }),
  input: (styles: StylesConfig) => ({
    ...styles,
    marginLeft: '-2px',
    marginTop: '0px',
    caretColor: '#999',
    boxShadow: 'none',
    border: 'none',
    maxHeight: '20px',
    minHeight: '20px',
    lineHeight: '12px'
  })
};

export const tagging = {
  control: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      borderColor: '#e6e6e6',
      boxShadow: 'none',
      transition: 'all 0.3s ease-in-out',
      '&:hover': {
        borderColor: '#e6e6e6'
      },
      '&:focus-within': {
        boxShadow: 'var(--widget-box-shadow-focused) !important',
        border: '1px solid #a4cafe !important'
      }
    };
  },
  input: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      marginLeft: '0px',
      caretColor: '#999',
      marginBottom: '0px',
      boxShadow: 'none',
      border: 'none'
    };
  },
  multiValue: (styles: StylesConfig) => {
    return {
      ...styles,
      maxHeight: 24,
      minHeight: 24,
      lineHeight: '17px',
      backgroundColor: 'hsl(0, 0%, 94%)',
      border: '1px solid hsl(0, 0%, 85%)'
    };
  }
};
