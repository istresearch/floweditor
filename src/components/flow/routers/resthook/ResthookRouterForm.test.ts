import { composeComponentTestUtils, mock, setMock } from 'testUtils';
import {
  getRouterFormProps,
  createResthookNode,
  createCallResthookAction,
  ResthookAsset
} from 'testUtils/assetCreators';
import * as utils from 'utils';
import ResthookRouterForm from 'components/flow/routers/resthook/ResthookRouterForm';

mock(utils, 'createUUID', utils.seededUUIDs());

const routerNode = createResthookNode(createCallResthookAction());
const { setup } = composeComponentTestUtils(ResthookRouterForm, getRouterFormProps(routerNode));

describe(ResthookRouterForm.name, () => {
  describe('render', () => {
    it('should render', () => {
      const { wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('updates', () => {
    it('should update and save', () => {
      const components = setup(true, { updateRouter: setMock() });
      const instance = components.instance as ResthookRouterForm;

      instance.handleResthookChanged([ResthookAsset]);
      expect(instance.state).toMatchSnapshot();
      instance.getButtons().primary.onClick();
      expect(components.props.updateRouter).toMatchCallSnapshot();
    });

    it('should cancel changes', () => {
      const components = setup(true, { updateRouter: setMock() });
      const instance = components.instance as ResthookRouterForm;

      instance.handleResthookChanged([ResthookAsset]);
      instance.getButtons().secondary.onClick();
      expect(components.props.updateRouter).not.toBeCalled();
    });

    it('coverts from other node types', () => {
      const components = setup(true, {
        updateRouter: setMock(),
        nodeSettings: {
          originalNode: { ui: { $merge: { type: null } } }
        }
      });

      const instance = components.instance as ResthookRouterForm;
      instance.handleResthookChanged([ResthookAsset]);
      instance.getButtons().primary.onClick();
      expect(components.props.updateRouter).toMatchCallSnapshot();
    });

    it('creates its own action uuid if necessary', () => {
      const components = setup(true, {
        updateRouter: setMock(),
        nodeSettings: {
          originalNode: {
            node: { $merge: { actions: [] } },
            ui: { $merge: { type: null } }
          }
        }
      });

      const instance = components.instance as ResthookRouterForm;

      instance.handleResthookChanged([ResthookAsset]);
      instance.getButtons().primary.onClick();
      expect(components.props.updateRouter).toMatchCallSnapshot();
    });

    it('validates before saving', () => {
      const components = setup(true, { updateRouter: setMock() });
      const instance = components.instance as ResthookRouterForm;

      instance.handleResthookChanged([]);
      instance.getButtons().primary.onClick();
      expect(components.props.updateRouter).not.toBeCalled();
    });
  });
});
