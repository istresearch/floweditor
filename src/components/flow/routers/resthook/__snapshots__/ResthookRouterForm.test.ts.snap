// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResthookRouterForm render should render 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="call_resthook"
  title="Call Zapier"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "aliases": Array [
          "split_by_resthook",
        ],
        "component": [Function],
        "description": "Call Zapier",
        "filter": "resthook",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
        ],
        "name": "Call Zapier",
        "type": "call_resthook",
        "visibility": Array [
          "messaging",
          "messaging_background",
          "voice",
        ],
      }
    }
    onChange={[MockFunction]}
  />
  <AssetSelector
    entry={
      Object {
        "value": Object {
          "resthook": "my-resthook",
        },
      }
    }
    name="Resthook"
    nameKey="resthook"
    onChange={[Function]}
    placeholder="Select the resthook to call"
    searchable={true}
    valueKey="resthook"
  />
  <div
    className="result_name"
  >
    <OptionalTextInput
      helpText={
        <Trans
          i18nKey="forms.result_name_help"
          values={
            Object {
              "resultFormat": "@results.result",
            }
          }
        >
          By naming the result, you can reference it later using [[resultFormat]]
        </Trans>
      }
      maxLength={64}
      name="Result Name"
      onChange={[Function]}
      toggleText="Save as.."
      value={
        Object {
          "value": "result",
        }
      }
    />
  </div>
</Dialog>
`;

exports[`ResthookRouterForm updates coverts from other node types 1`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [
        Object {
          "resthook": "new-resthook",
          "result_name": "result",
          "type": "call_resthook",
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
        },
      ],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        },
        Object {
          "destination_uuid": null,
          "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "Success",
            ],
            "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "type": "has_only_text",
            "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
            "name": "Success",
            "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
          },
          Object {
            "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            "name": "Failure",
            "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
          },
        ],
        "default_category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
        "operand": "@results.result.category",
        "type": "switch",
      },
      "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_resthook",
    },
  },
]
`;

exports[`ResthookRouterForm updates creates its own action uuid if necessary 1`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [
        Object {
          "resthook": "new-resthook",
          "result_name": "Result",
          "type": "call_resthook",
          "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
        },
      ],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
        },
        Object {
          "destination_uuid": null,
          "uuid": "e0bc3776-792f-49d2-b4be-48be2cf8159a",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "Success",
            ],
            "category_uuid": "6c22884b-0e35-4206-982e-18320691eda9",
            "type": "has_only_text",
            "uuid": "0c8c9239-9b59-473a-a229-d8a9693be270",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
            "name": "Success",
            "uuid": "6c22884b-0e35-4206-982e-18320691eda9",
          },
          Object {
            "exit_uuid": "e0bc3776-792f-49d2-b4be-48be2cf8159a",
            "name": "Failure",
            "uuid": "a62e8284-49f8-4292-adbc-d5f6607f85c5",
          },
        ],
        "default_category_uuid": "a62e8284-49f8-4292-adbc-d5f6607f85c5",
        "operand": "@results.result.category",
        "type": "switch",
      },
      "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_resthook",
    },
  },
]
`;

exports[`ResthookRouterForm updates should update and save 1`] = `
Object {
  "resthook": Object {
    "validationFailures": Array [],
    "value": Object {
      "name": "new-resthook",
      "resthook": "new-resthook",
      "type": "resthook",
    },
  },
  "resultName": Object {
    "value": "result",
  },
  "valid": true,
}
`;

exports[`ResthookRouterForm updates should update and save 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [
        Object {
          "resthook": "new-resthook",
          "result_name": "result",
          "type": "call_resthook",
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
        },
      ],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        },
        Object {
          "destination_uuid": null,
          "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "Success",
            ],
            "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "type": "has_only_text",
            "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
            "name": "Success",
            "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
          },
          Object {
            "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            "name": "Failure",
            "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
          },
        ],
        "default_category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
        "operand": "@results.result.category",
        "type": "switch",
      },
      "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_resthook",
    },
  },
]
`;
